experimentation/tools/sorald/handled_rules.py

```diff
@@ -0,0 +1,203 @@
+        required=True,
+    )
+    return parser.parse_args(args)
+
+
+def generate_handled_rules(
+    output_file: pathlib.Path,
+    template: str,
+) -> None:
+    spoon_output = get_output_from_spoon()
+    parse_spoon_output(spoon_output)
+    rendered_content = jinja2.Template(template).render(
+        bugs=BUG,
+        bugs_detail=BUG_DETAIL,
+        code_smells=CODE_SMELL,
+        code_smells_detail=CODE_SMELL_DETAIL,
+        vulnerabilities=VULNERABILITY,
+        vulnerabilities_detail=VULNERABILITY_DETAIL,
+    )
+    output_file.write_text(rendered_content, encoding=ENCODING)
+
+
+def get_output_from_spoon() -> List[SoraldProcessorInformation]:
+    command = f"java -cp {CLASSPATH} {PATH_TO_JAVA_SCRIPT} {PATH_TO_PROCESSOR_PACKAGE}"
+    pipe = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    output, error = pipe.communicate()
+    if error:
+        raise Exception(error.decode(ENCODING))
+    return [SoraldProcessorInformation(**processor) for processor in json.loads(output.decode(ENCODING))]
+
+
+def parse_spoon_output(spoon_output: List[SoraldProcessorInformation]) -> None:
+    for processor in spoon_output:
+        classify(
+            sonar_metadata.get_rule_metadata(processor.rule_key),
+            processor.rule_key,
+            processor.repair_description,
+        )
+
+
+def classify(metadata: Dict[str, Union[str, List[str]]], rule_key: int, repair_description: str) -> None:
+    list_information = ViolationList(
+        title=metadata[jsonkeys.SONAR_METADATA.TITLE],
+        link_to_detail=get_link_to_detail(metadata[jsonkeys.SONAR_METADATA.TITLE], rule_key),
+        sonar_url=get_sonar_link(metadata[jsonkeys.SONAR_METADATA.RULE_SPECIFICATION]),
+        sonar_url_text=f"Sonar Rule {rule_key}",
+    )
+    detail_information = ViolationDetail(
+        title=metadata[jsonkeys.SONAR_METADATA.TITLE],
+        sonar_url=get_sonar_link(metadata[jsonkeys.SONAR_METADATA.RULE_SPECIFICATION]),
+        sonar_url_text=f"Sonar Rule {rule_key}",
+        repair_description=repair_description,
+    )
+
+    violation_type = metadata[jsonkeys.SONAR_METADATA.TYPE]
+    if violation_type == sonar_metadata.VIOLATION_TYPE.BUG:
+        BUG.append(list_information)
+        BUG_DETAIL.append(detail_information)
+
+    elif violation_type == sonar_metadata.VIOLATION_TYPE.CODE_SMELL:
+        CODE_SMELL.append(list_information)
+        CODE_SMELL_DETAIL.append(detail_information)
+
+    elif violation_type == sonar_metadata.VIOLATION_TYPE.VULNERABILITY:
+        VULNERABILITY.append(list_information)
+        VULNERABILITY_DETAIL.append(detail_information)
+
+    else:
+        raise Exception(f"New rule violation type, {violation_type}, encountered.")
+
+
+def get_link_to_detail(title: str, rule_key: int) -> str:
+    link = title.lower()
+    link = re.sub(r"\s", "-", link)
+    link = re.sub(r"[^A-Za-z0-9_-]", "", link)
+    return f"#{link}-sonar-rule-{rule_key}"
```

Comment:
This function is to create in-page links for the repairs. I wrote this based on general observation. I am not sure about the exact conventions of how HTML links headers in pages.

Reply:
The links are generated by GitHub markdown. It's typically just the content of the heading in lowercase, with spaces replaced by dashes. For example `# Hello there` can be linked to as `[Go to hello there](#hello-there)`. Some special symbols are removed, but I don't know off the top of my head which ones those are.

Anyway, the gist of it is that a link to a given section heading in the Markdown file should be generated from the text in the heading.

Reply:
> Some special symbols are removed, but I don't know off the top of my head which ones those are.

Apart from `_` and `-`, I have removed everything else. If a special case arises, I will handle it then.

Reply:
Okay. So you are suggesting that I pass the entire heading text instead of passing the title and rule key separately?

Also, I didn't know about `remove_special_chars`. I will use that instead in this function.

Reply:
> Okay. So you are suggesting that I pass the entire heading text instead of passing the title and rule key separately?

Yes, because that's what's used to generate the link. You should be able to pass an arbitrary heading to it. It's also easier to understand as there are no magic words like in the current function.

`remove_special_chars` isn't a defined function, I was just to lazy to copy-paste your regex.

Reply:
Edited, that ought to work.